File: hadoop/src/MutualFriendZip.java
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FileStatus;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.LongWritable;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.Set;

/**
 * Created by xiezebin on 2/11/16.
 */
public class MutualFriendZip extends MutualFriend
{
    private final static String USAGE = "Usage: MutualFriendZip <in_adj> <in_data> <out> <user id,user id>";
    private final static String TEMP_FOLDER = "/zxx140430/output/temp";


    // in-memory join
    public static class MapForStep2
            extends Mapper<LongWritable, Text, Text, Text>
    {
        String obFriendPair;
        Set<String> obMutualUIDSet;


        public void map(LongWritable key, Text value, Context context
        ) throws IOException, InterruptedException
        {
            // case 1: no mutual friend, just return    todo emit one empty record
            if (obMutualUIDSet.isEmpty())
            {
                return;
            }

            String[] loLineOfData = value.toString().split(",");
            String loCurrentUID = loLineOfData[0];
            // case 2: current user is not mutual friend, just return
            if (!obMutualUIDSet.contains(loCurrentUID))
            {
                return;
            }

            // emit data
            String loFriendName = loLineOfData[1] + " " + loLineOfData[2];
            String loValueOut = loLineOfData[6];        // zip
            context.write(new Text(obFriendPair), new Text(loFriendName + ":" + loValueOut));
        }

        protected void setup(Context context) throws IOException, InterruptedException
        {
            super.setup(context);

            Configuration conf = context.getConfiguration();
            String loPathString = conf.get(DRIVER_TO_MAPPER_KEY);
            Path loPath = new Path("hdfs://cshadoop1" + loPathString);  // Location of file in HDFS

            obMutualUIDSet = new HashSet<>();

            FileSystem fs = FileSystem.get(conf);
            FileStatus[] fss = fs.listStatus(loPath);
            for (FileStatus status : fss)                      // read all files under path
            {
                Path pt = status.getPath();
                BufferedReader br=new BufferedReader(new InputStreamReader(fs.open(pt)));
                String line = br.readLine();
                while (line != null)
                {
                    String[] loCacheDataSp = line.split("\\t");
                    if(loCacheDataSp.length > 1)
                    {
                        obFriendPair = loCacheDataSp[0];

                        // save mutual friend id
                        String[] loMutualUID = loCacheDataSp[1].split(",");
                        for (int i = 0; i < loMutualUID.length; i++)
                        {
                            obMutualUIDSet.add(loMutualUID[i]);
                        }
                    }
                    line=br.readLine();
                }
            }
        }
    }

    public static class ReduceForStep2
            extends Reducer<Text, Text, Text, Text>
    {
        public void reduce(Text key, Iterable<Text> values, Context context
        ) throws IOException, InterruptedException
        {
            StringBuilder loBuilder = new StringBuilder();
            for (Text lpMutualZip : values)
            {
                loBuilder.append(",");
                loBuilder.append(lpMutualZip);
            }
            if (loBuilder.length() > 0)
            {
                loBuilder.deleteCharAt(0);
            }

            context.write(new Text(key), new Text(loBuilder.toString()));
        }
    }


    // Driver program
    public static void main(String[] args) throws Exception
    {
        String[] otherArgs = getOtherArgs(args, 4, USAGE);

        // ==== EXEC first step =====================
        String loTargetUID = otherArgs[3];
        Job loJobStep1 = setupJob(Map.class, Reduce.class, otherArgs[0], TEMP_FOLDER, loTargetUID);

        // ==== EXEC second step ====================
        if (loJobStep1.waitForCompletion(true))
        {
            Job loJobStep2 = setupJob(MapForStep2.class, ReduceForStep2.class, otherArgs[1], otherArgs[2], TEMP_FOLDER);

            System.exit(loJobStep2.waitForCompletion(true) ? 0 : 1);
        }
        else
        {
            System.exit(1);
        }

    }
}

File: sparkcodecluster/src/main/scala/MutualFriendZip.scala
/**
 * Created by xiezebin on 3/7/16.
 */

import org.apache.spark.{SparkContext, SparkConf}
import MutualFriend.reduceProcess

object MutualFriendZip {

  val INPUT_DIR = "hdfs://cshadoop1/zxx140430/input/hw1-Adj/soc-LiveJournal1Adj.txt"
  val INPUT_INFO_DIR = "hdfs://cshadoop1/zxx140430/input/hw1-userdata"
  val OUTPUT_DIR = "hdfs://cshadoop1/zxx140430/output/"

  def mapProcessZip(value:String, mutualSet:Set[String]): String = {
    // case 1: no mutual friend, just return    todo emit one empty record
    if (mutualSet.isEmpty)
    {
      return ""
    }

    val loLineOfData = value.split(",")
    val loCurrentUID = loLineOfData(0)
    // case 2: current user is not mutual friend, just return
    if (!mutualSet.contains(loCurrentUID))
    {
      return ""
    }

    // emit data
    val loFriendName = loLineOfData(1) + " " + loLineOfData(2)
    val loValueOut = loLineOfData(6)              // zip
    loFriendName + ":" + loValueOut
  }

  def reduceProcessZip(args:(String, Iterable[String])): String = {
    val loBuilder = new StringBuilder()
    args._2.foreach(lpMutualZip =>
    {
      loBuilder.append(",")
      loBuilder.append(lpMutualZip)
    })
    if (loBuilder.nonEmpty)
    {
      loBuilder.deleteCharAt(0)
    }

    args._1 + "\t" + loBuilder.toString()
  }

  def main(args: Array[String]) {
    val conf = new SparkConf().setAppName("Friend Recommendation Project")
    conf.set("spark.hadoop.validateOutputSpecs", "false")
    val sc = new SparkContext(conf)

    var loTargetUIDSet = Set[String]()
    if (args.length > 0)
    {
      args(0).split(",").foreach(uid => loTargetUIDSet += uid)
    }
    var targetBc = sc.broadcast(loTargetUIDSet)

    val logData = sc.textFile(INPUT_DIR)
    val loMapStepOneResult = logData.filter(line => targetBc.value.contains(line.split("\\t")(0)))
    val loReduceStepOneResult = loMapStepOneResult.groupBy(line => args(0)).map(reduceProcess)


    // create mutual set for step two
    val loMutualArray = loReduceStepOneResult.collect()(0).split("\\t")(1).split(",")
    var loMutualSet = Set[String]()
    loMutualArray.foreach(friend => loMutualSet += friend)
    targetBc = sc.broadcast(loMutualSet)


    val loUserData = sc.textFile(INPUT_INFO_DIR)
    val loFilterResult = loUserData.filter(line => targetBc.value.contains(line.split(",")(0)))
    val loFindZipResult = loFilterResult.map(line => mapProcessZip(line, targetBc.value))
    val loShuffleResult = loFindZipResult.groupBy(line => args(0))
    val loFinalResult = loShuffleResult.map(reduceProcessZip)

    loFinalResult.saveAsTextFile(OUTPUT_DIR + "test")
  }
}
